core/src/main/java/jenkins/security/apitoken/ApiTokenStats.java

```diff
@@ -0,0 +1,262 @@
+import hudson.model.listeners.SaveableListener;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.io.File;
+import java.io.IOException;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public class ApiTokenStats implements Saveable {
+    private static final Logger LOGGER = Logger.getLogger(ApiTokenStats.class.getName());
+    
+    /**
+     * Normally a user will not have more 2-3 tokens at a time, 
+     * so there is no need to store a map here
+     */
+    private List<SingleTokenStats> tokenStats;
+    
+    private transient File parent;
+    
+    public ApiTokenStats() {
+        this.init();
+    }
+    
+    private ApiTokenStats readResolve() {
+        this.init();
+        return this;
+    }
+    
+    private void init() {
+        if (this.tokenStats == null) {
+            this.tokenStats = new ArrayList<>();
+        } else {
+            keepLastUpdatedDuplicated();
+        }
+    }
+    
+    private void keepLastUpdatedDuplicated() {
+        Map<String, SingleTokenStats> temp = new HashMap<>();
+        this.tokenStats.forEach(candidate -> {
+            SingleTokenStats current = temp.get(candidate.tokenUuid);
+            if (current == null) {
+                temp.put(candidate.tokenUuid, candidate);
+            } else {
+                int comparison = SingleTokenStats.COMP_BY_LAST_USE_THEN_COUNTER.compare(current, candidate);
+                if (comparison < 0) {
+                    temp.put(candidate.tokenUuid, candidate);
+                }
+            }
+        });
+        
+        this.tokenStats = new ArrayList<>(temp.values());
+    }
+    
+    public void setParent(@Nonnull File parent) {
+        this.parent = parent;
+    }
+    
+    private boolean areStatsDisabled(){
+        return !ApiTokenPropertyConfiguration.get().isUsageStatisticsEnabled();
+    }
+    
+    /**
+     * Will trigger the save if there is some modification
+     */
+    public synchronized void removeId(@Nonnull String tokenUuid) {
+        if(areStatsDisabled()){
+            return;
+        }
+        
+        boolean tokenRemoved = tokenStats.removeIf(s -> s.tokenUuid.equals(tokenUuid));
+        if (tokenRemoved) {
+            save();
+        }
+    }
+    
+    /**
+     * Will trigger the save
+     */
+    public synchronized @Nonnull SingleTokenStats updateUsageForId(@Nonnull String tokenUuid) {
+        if(areStatsDisabled()){
+            return new SingleTokenStats(tokenUuid);
+        }
+        
+        SingleTokenStats stats = tokenStats.stream()
+                .filter(s -> s.tokenUuid.equals(tokenUuid))
+                .findFirst()
+                .orElse(null);
+        
+        if (stats == null) {
```

Comment:
I feel like using `Optional` and `null` at the same time is a bit of an anti-pattern.

Reply:
I changed the code, it looks like better, 👍